go/store/nbs/generational_chunk_store.go

```diff
@@ -0,0 +1,302 @@
+	return gcs.oldGen
+}
+
+// Get the Chunk for the value of the hash in the store. If the hash is absent from the store EmptyChunk is returned.
+func (gcs *GenerationalNBS) Get(ctx context.Context, h hash.Hash) (chunks.Chunk, error) {
+	c, err := gcs.oldGen.Get(ctx, h)
+
+	if err != nil {
+		return chunks.EmptyChunk, err
+	}
+
+	if c.IsEmpty() {
+		return gcs.newGen.Get(ctx, h)
+	}
+
+	return c, nil
+}
+
+// GetMany gets the Chunks with |hashes| from the store. On return, |foundChunks| will have been fully sent all chunks
+// which have been found. Any non-present chunks will silently be ignored.
+func (gcs *GenerationalNBS) GetMany(ctx context.Context, hashes hash.HashSet, found func(*chunks.Chunk)) error {
+	notInOldGen := hashes.Copy()
+	err := gcs.oldGen.GetMany(ctx, hashes, func(chunk *chunks.Chunk) {
+		delete(notInOldGen, chunk.Hash())
+		found(chunk)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if len(notInOldGen) == 0 {
+		return nil
+	}
+
+	return gcs.newGen.GetMany(ctx, notInOldGen, found)
+}
+
+func (gcs *GenerationalNBS) GetManyCompressed(ctx context.Context, hashes hash.HashSet, found func(CompressedChunk)) error {
+	notInOldGen := hashes.Copy()
+	err := gcs.oldGen.GetManyCompressed(ctx, hashes, func(chunk CompressedChunk) {
+		delete(notInOldGen, chunk.Hash())
+		found(chunk)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if len(notInOldGen) == 0 {
+		return nil
+	}
+
+	return gcs.newGen.GetManyCompressed(ctx, notInOldGen, found)
+}
+
+// Returns true iff the value at the address |h| is contained in the
+// store
+func (gcs *GenerationalNBS) Has(ctx context.Context, h hash.Hash) (bool, error) {
+	has, err := gcs.oldGen.Has(ctx, h)
+
+	if err != nil {
+		return false, err
+	}
+
+	if has {
+		return true, nil
+	}
+
+	return gcs.newGen.Has(ctx, h)
+}
+
+// Returns a new HashSet containing any members of |hashes| that are
+// absent from the store.
+func (gcs *GenerationalNBS) HasMany(ctx context.Context, hashes hash.HashSet) (absent hash.HashSet, err error) {
+	notInOldGen, err := gcs.oldGen.HasMany(ctx, hashes)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(notInOldGen) == 0 {
+		return notInOldGen, nil
+	}
+
+	return gcs.newGen.HasMany(ctx, notInOldGen)
+}
+
+// Put caches c in the ChunkSource. Upon return, c must be visible to
+// subsequent Get and Has calls, but must not be persistent until a call
+// to Flush(). Put may be called concurrently with other calls to Put(),
+// Get(), GetMany(), Has() and HasMany().
+func (gcs *GenerationalNBS) Put(ctx context.Context, c chunks.Chunk) error {
+	return gcs.newGen.Put(ctx, c)
+}
+
+// Returns the NomsVersion with which this ChunkSource is compatible.
+func (gcs *GenerationalNBS) Version() string {
+	return gcs.newGen.Version()
+}
+
+// Rebase brings this ChunkStore into sync with the persistent storage's
+// current root.
+func (gcs *GenerationalNBS) Rebase(ctx context.Context) error {
+	oErr := gcs.oldGen.Rebase(ctx)
+	nErr := gcs.newGen.Rebase(ctx)
```

Comment:
How do we think about the order here? We opened `new` before `old` previously, and here we're rebasing `old` and then `new`. What's the actual order we want? What conditions and invariants can we reason about so we can be certain we have a consistent unified view of old+new after we open or Rebase?